### PR TITLE
Lazy load parquet

### PR DESCRIPTION
**Pull request by dispanser**
@houqp as requested, the initial PoC of using `TableProvider` trait instead of explicit `MemTable`, to enable direct queries on the source storage via datafusion.

### CODE REVIEW

**Comment by dispanser** [comment]
@houqp : made a config switch, `lazy: bool`, probably not a good name.

Tests are missing, but I'd like to hear your thoughts before moving on and adding delta / csv.

**Comment by houqp** [comment]
Thanks @dispanser , I am on vacation this week, but I am going to find some time to play with your branch today or tomorrow to better understand the gaps we have in datafusion :)

**Comment by dispanser** [comment]
maybe saves you some effort during your vacation, the specific problem with `blogs.parquet` is:

```
thread 'table::parquet::tests::load_partitions' panicked at 'index out of bounds: the len is 2 but the index is 2', /home/pi/.cargo/registry/src/github.com-1ecc6299db9ec823/datafusion-5.0.0/src/physical_plan/parquet.rs:226:25
```
which is part of

```rust
                    for (i, cnt) in columns_null_counts.enumerate() {
                        null_counts[i] += cnt
                    }

```

Which results from the parquet reader allocating `null_counts` a `vec` of size two because the schema has two (top-level) field entries, but the number of null counts from  `columns_null_counts.enumerate()` is three, because it enumerates nested columns.

**Comment by houqp** [comment]
Thanks @dispanser for the notes, it definitely saved me a lot of time :) I think this is a valid bug we should fix in datafusion. Datafusion today doesn't support querying nested columns, but it should be able to load dataset with nested columns without error and let user query non-nested columns just like how it works for memory tables. We can work on the fix once https://github.com/apache/arrow-datafusion/pull/932 is merged.

For now, I think it's fine to ignore that test error in CI and merge in the current design. After we get it fixed in Datafusion, we can reenable the test that does lazy loading with `blogs.parquet`.

In the long run, we should pick up https://github.com/apache/arrow-datafusion/pull/628 to add nested field query support, then update the code base to support nested column stats.

File: columnq/src/table/parquet.rs [hunk 1]
```diff
@@ -1,78 +1,89 @@
...
-                .map_err(ColumnQError::parquet_record_reader)?;
-
-            let batch_schema = arrow_reader.get_schema().map_err(|_| {
-                ColumnQError::LoadParquet("failed to load schema from partition".to_string())
+pub async fn to_mem_table(t: &TableSource) -> Result<Arc<dyn TableProvider>, ColumnQError> {
+    let lazy = match &t.option {
+        Some(TableLoadOption::parquet { lazy }) => lazy.unwrap_or(false),
+        _ => false,
+    };
+
```
Comment (houqp) [inline columnq/src/table/parquet.rs:21]: would be cleaner to wrap `to_mem_table` with a wrapper function that matches on table option:

```rust
pub async fn to_datafusion_table() -> Result<Arc<dyn TableProvider>, ColumnQError> {
    match t.option {
        Some(TableLoadOption::parquet { lazy }) if lazy == Some(true) => Ok(Arc::new(ParquetTable::try_new(t.parsed_uri()?, 4).map_err(|err| {
            ColumnQError::LoadParquet(format!("failed to load parquet: '{}'", err.to_string()))
        })?)),
        _ => to_mem_table(t),
    }
}
```

File: columnq/src/table/mod.rs [hunk 2]
```diff
@@ -137,7 +140,9 @@ pub enum TableLoadOption {
     },
     csv(TableOptionCsv),
     ndjson {},
-    parquet {},
+    parquet {
+        lazy: Option<bool>,
```
Comment (houqp) [inline columnq/src/table/mod.rs:144]: We could use the same defaults pattern in the csv option struct to get rid of the option here, which should simplify downstream code by avoiding option match on every access.

**Review by houqp — COMMENTED** [review]
Thanks @dispanser ! The general direction looks great to me :)

File: columnq/src/table/mod.rs (hunk 2 above)
Comment (dispanser) [reply columnq/src/table/mod.rs]: Modelled `TableOptionParquet` oafter `TableOptionCsv`, hope that fits the bill now.

Also renamed the parameter and inverted the meaning to `use_memory_table`, defaulting to true to maintain the current roapi behavior as the default. IMHO, this name better conveys the purpose of the configuration option than `lazy`.

**Comment by dispanser** [comment]
I'd like to add the same functionality to `DeltaTable`. Would you prefer a separate PR or just add it here?

As for csv, I'm not even sure it makes a lot of sense. Any CSV source that's big enough to not fit in memory is probably too big anyway. IIUC correctly, with the absence of file stats and partition pruning, every query would have to read all of the csv anyway, so it's better to just pay that price once.

**Comment by houqp** [comment]
> I'd like to add the same functionality to DeltaTable. Would you prefer a separate PR or just add it here?

Up to you :) Just pick the route with the least resistance.

File: columnq/src/table/parquet.rs [hunk 3]
```diff
@@ -1,19 +1,39 @@
...
-use crate::error::ColumnQError;
-use crate::table::TableSource;
+pub async fn to_datafusion_table(t: &TableSource) -> Result<Arc<dyn TableProvider>, ColumnQError> {
+    let opt = t
+        .option
+        .clone()
+        .unwrap_or_else(|| TableLoadOption::parquet(TableOptionParquet::default()));
+    let TableOptionParquet { use_memory_table } = opt.as_parquet()?;
+
+    println!("{}", use_memory_table);
```
Comment (houqp) [inline columnq/src/table/parquet.rs:23]: left over debug code?

**Comment by dispanser** [comment]
> > I'd like to add the same functionality to DeltaTable. Would you prefer a separate PR or just add it here?
> 
> Up to you :) Just pick the route with the least resistance.

I prefer a separate PR. From my perspective, we need:
- a test (I'll probably use a flattened version of `blogs.parquet` for this
- readme updates

**Comment by houqp** [comment]
sounds good, @dispanser is this ready to merge?

**Comment by dispanser** [comment]
@houqp : I added a test, and consider this done now. I noticed that the documentation is in a separate repository, I'll look into that to add documentation (and caveats) for the new feature.

**Comment by dispanser** [comment]
@houqp : One more thing that came to my mind: would it be good to bail out if the user attempts to use a storage option that is not directly supported by datafusion and thus requires a `MemTable`?

**Review by houqp — APPROVED** [review]

**Comment by houqp** [comment]
@dispanser looks like there is a conflict with the latest main branch that's blocking the merge.

> One more thing that came to my mind: would it be good to bail out if the user attempts to use a storage option that is not directly supported by datafusion and thus requires a MemTable?

Good call, i believe the code you have should result in an early error return from `ParquetTable::try_new` right?

**Comment by dispanser** [comment]
@houqp : I will try to cover that in the documentation a bit. there are multiple caveats with this solution  (nested schemas not supported, only file-system, ...) and only some of them can be detected easily beforehand anyway.

If you're happy with the state of this, feel free to merge, I'll create a follow-up PR for delta, and then make the doc changes in one pass if you don't mind.

**Comment by houqp** [comment]
@dispanser the change looks great to me, but I couldn't merge this PR because there is a conflict :)

**Comment by dispanser** [comment]
@houqp : should be fixed now. I don't think I can see that from the github PR, because I don't have the "merge" button so it doesn't helpfully tell me that merge is currently impossible.

**Pull request merged**